### PR TITLE
Log to k8s events

### PR DESCRIPTION
**Pull request by nolar**
> Issue : #136 

_As a logical continuation of #125:_

## Description

Kopf is not a K8s API client library. Therefore, it should not provide any means of talking to the K8s API. The operators developers should use the K8s API client libraries of their choice. This also applies to K8s-event posting.

However, Kopf provides few routines for k8s-event posting: `kopf.event()`, `kopf.info()`, `kopf.warn()`, `kopf.exception()`. This is highly useful for logging not only to the logs (stdout/stderr), but also to K8s itself.

To deal with this dilemma, this PR changes the way how Kopf deals with this specific field — logging to k8s events:

Instead of providing the methods to explicitly post everything in addition to regular logging (thus duplicating the messages and code), everything logged to the per-object logger is now posted to K8s events transparently.

Was:

```python
@kopf.on.create('zalando.org', 'v1', 'kopfexamples')
def create_fn(logger, **kwargs):
    child = _create_child()
    kopf.info("Created a child: %s".format(child.id))
    logger.info("Created a child: %s", child.id)
```

Now:

```python
@kopf.on.create('zalando.org', 'v1', 'kopfexamples')
def create_fn(logger, **kwargs):
    child = _create_child()
    logger.info("Created a child: %s", child.id)
```

In the latter case, an k8s-event will be created too.

Only the log messages with level INFO+ will be posted. The DEBUG messages will **not** be posted (but will be logged as usually).

Only the messages logged to the per-object logger are posted. The rest of the `logging` machinery is **not** affected.

The existing k8s-event-posting routines remain in place. They continue to post the events as usually, attached to the specified object (their first positional argument).

All of the internal Kopf-originated events are now also posted that way — via the per-object logger (some were lowered to the DEBUG level to reduce noise).

## Types of Changes

- New feature (non-breaking change which adds functionality)
- Refactor/improvements

## Review
_List of tasks the reviewer must do to review the PR_
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: kopf/engines/logging.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    A logger/adapter to carry the object identifiers for formatting.
+
+    The identifiers are then used both for formatting the per-object messages
+    in `ObjectPrefixingFormatter`, and when posting the per-object k8s-events.
+
+    Constructed in event handling of each individual object.
+
+    The internal structure is made the same as an object reference in K8s API,
+    but can change over time to anything needed for our internal purposes.
+    However, the as little information should be carried as possible,
```
Comment (thilp) [inline kopf/engines/logging.py]:
```suggestion
    However, as little information should be carried as possible,
```

File: tests/posting/test_log2k8s.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+OBJ1 = {'apiVersion': 'group1/version1', 'kind': 'Kind1',
+        'metadata': {'uid': 'uid1', 'name': 'name1', 'namespace': 'ns1'}}
+REF1 = {'apiVersion': 'group1/version1', 'kind': 'Kind1',
+        'uid': 'uid1', 'name': 'name1', 'namespace': 'ns1'}
+
+
+@pytest.fixture()
+def logstream(caplog):
+    """ Prefixing is done at the final output. We have to intercept it. """
```
Comment (thilp) [inline tests/posting/test_log2k8s.py]: This fixture is almost identical to the one in `test_cause_logging.py`, except it sets `logger.handlers` back to its old value after the test (whereas the other fixture sets `logger.handlers` to the empty list afterwards). Why this difference?

File: kopf/__init__.py [hunk 3]
```diff
@@ -20,7 +20,10 @@
     LOGLEVEL_ERROR,
     LOGLEVEL_CRITICAL,
     EventsConfig,
-    WorkersConfig
+    WorkersConfig,
+)
+from kopf.engines.logging import (
+    ObjectLogger,
```
Comment (thilp) [inline kopf/__init__.py]: Why export this type at such a high level? It is not used from the root `kopf` module internally and, if I understand correctly, the only interactions the user could/should have with it are:
- use the instance already provided as the `logger` parameter in handlers (so no need to call the constructor),
- in type annotations, where encouraging users to rather use the _Logger_  or _LoggerAdapter_ interfaces would give Kopf more freedom for future refactorings.

Comment (nolar) [reply kopf/__init__.py]: You are right, let's not expose it to the public interface until there is a use-case. It is highly doubtful if anyone will instantiate their own per-object loggers at all. And it depend on the event-posting asyncio queue instance, which is surely not available to the users (it is part of Kopf's internals).

I will remove it from here.

File: tests/posting/test_log2k8s.py (hunk 2 above)
Comment (nolar) [reply tests/posting/test_log2k8s.py]: Overlooked that. Fixed. Now, the fixture is in top-level `conftest.py`.

File: kopf/__init__.py (hunk 3 above)
Comment (nolar) [reply kopf/__init__.py]: Removed.

**Pull request merged**